multi/stm32l4-multi/libmulti/libdma.c

```diff
@@ -11,205 +11,452 @@
+
+	return 0;
 }
 
 
-static int _has_channel_finished(const void *maddr, volatile unsigned int *channel_base)
+static int libdma_hasChannelFinished(volatile unsigned int *channelBase)
 {
-	return (maddr != NULL && *(channel_base + cndtr) > 0);
+	return (*(channelBase + cndtr) > 0) ? 1 : 0;
 }
 
 
-static int _transfer(int dma, int rx_channel, int tx_channel, void *rx_maddr, const void *tx_maddr, size_t len)
+static int libdma_transferHelper(int dma, int rxChannel, int txChannel, void *rxMAddr, const void *txMAddr, size_t len)
 {
 	/* Empirically chosen value to avoid mutex+cond overhead for short transactions */
 	int use_interrupts = len > 24;
+	int interrupts_flags = (use_interrupts == 0) ? 0 : DMA_TCIE_FLAG;
+	volatile int rxDone = 0;
 
-	volatile unsigned int *rx_channel_base = dma_common[dma].base + 2 + 5 * rx_channel;
-	volatile unsigned int *tx_channel_base = dma_common[dma].base + 2 + 5 * tx_channel;
+	volatile unsigned int *rxChannel_base = dma_common[dma].base + 2 + (5 * rxChannel);
+	volatile unsigned int *txChannel_base = dma_common[dma].base + 2 + (5 * txChannel);
+
+	/* Only one request may be issued on one channel at one time. */
+	if ((dma_transfers[dma][rxChannel].type != dma_transferNull) ||
+		(dma_transfers[dma][txChannel].type != dma_transferNull)) {
+		return -EINVAL;
+	}
 
-	if (rx_maddr != NULL)
-		_prepare_transfer(rx_channel_base, rx_maddr, len, use_interrupts);
+	dma_transfers[dma][rxChannel].type = dma_transferOnce;
+	dma_transfers[dma][rxChannel].once.doneFlag = &rxDone;
+	dma_transfers[dma][rxChannel].once.channelBase = rxChannel_base;
+	libdma_prepareTransfer(rxChannel_base, rxMAddr, len, interrupts_flags);
 
-	if (tx_maddr != NULL)
-		/* When doing rw transfer, avoid unnecessary interrupt handling and condSignal()
-		by waiting only for RX transfer completion, ignoring TX */
-		_prepare_transfer(tx_channel_base, (void *)tx_maddr, len, use_interrupts && rx_maddr == NULL);
+	dma_transfers[dma][txChannel].type = dma_transferOnce;
+	/* When doing rw transfer, avoid unnecessary interrupt handling and condSignal()
+	by waiting only for RX transfer completion, ignoring TX */
+	libdma_prepareTransfer(txChannel_base, (void *)txMAddr, len, 0);
 
-	if (use_interrupts) {
+	if (use_interrupts != 0) {
 		mutexLock(dma_common[dma].irqLock);
-		while (_has_channel_finished(rx_maddr, rx_channel_base) || _has_channel_finished(tx_maddr, tx_channel_base))
-			condWait(dma_common[dma].cond, dma_common[dma].irqLock, 1);
+		while ((rxDone == 0) || (libdma_hasChannelFinished(txChannel_base) == 0)) {
+			condWait(dma_common[dma].cond, dma_common[dma].irqLock, DMA_TIMEOUT_1S);
+		}
 		mutexUnlock(dma_common[dma].irqLock);
 	}
 	else {
-		while (_has_channel_finished(rx_maddr, rx_channel_base) || _has_channel_finished(tx_maddr, tx_channel_base))
+		while ((libdma_hasChannelFinished(rxChannel_base) == 0) || (libdma_hasChannelFinished(txChannel_base) == 0))
```

Comment:
Same as above, because of the reversed logic of libdma_hasChannelFinished, we wait for rx and tx channels to both be **not finished** instead of waiting for both channels to finish.